server/src/main/java/org/elasticsearch/action/bulk/BulkProcessor2.java

```diff
@@ -295,6 +321,44 @@ private BulkProcessor2 add(DocWriteRequest<?> request) throws EsRejectedExecutio
         return this;
     }
 
+    private BulkProcessor2 addWithBackpressure(DocWriteRequest<?> request, Supplier<Boolean> shouldAbort)
+        throws EsRejectedExecutionException {
+        /*
+         * We want this method to block until the bulkProcessor accepts the request. Otherwise, the subsequent calls to this method will
+         * likely continue rejecting. BulkProcessor2.add does not have the ability to exert backpressure.
+         * So we have to catch the EsRejectedExecutionException that is thrown when it already has too many bytes in flight. We then
+         * wait until some bulk has been completed, reducing the amount of data in flight and (probably) making room for this request.
+         */
+        boolean successfullyAdded = false;
+        while (successfullyAdded == false) {
+            if (shouldAbort.get()) {
+                throw new EsRejectedExecutionException("Rejecting request because shouldAbort is now true");
```

Comment:
Nit: mentioning `shouldAbort` is a little weird here, should we reformat the message to just say that the caller has cancelled the bulk operation?